sources/scraping/runner.py

```diff
@@ -0,0 +1,183 @@
+import threading
+import typing as t
+
+import dlt
+from dlt.common import logger
+from pydispatch import dispatcher  # type: ignore
+
+from scrapy import signals, Item, Spider  # type: ignore
+from scrapy.crawler import CrawlerProcess  # type: ignore
+from scrapy.exceptions import CloseSpider  # type: ignore
+
+from .types import AnyDict, Runnable, P
+from .queue import ScrapingQueue
+
+T = t.TypeVar("T")
+
+
+class Signals:
+    def __init__(self, pipeline_name: str, queue: ScrapingQueue[T]) -> None:
+        self.queue = queue
+        self.pipeline_name = pipeline_name
+
+    def on_item_scraped(self, item: Item) -> None:
+        if not self.queue.is_closed:
+            self.queue.put(item)
+        else:
+            logger.info(
+                "Queue is closed ",
+                extra={"pipeline_name": self.pipeline_name},
+            )
+            raise CloseSpider("Queue is closed")
```

Comment:
looks like you cannot raise this to close spider:
https://github.com/scrapy/scrapy/issues/3435

we'd need to have something else. we may just schedule to stop crawl/engine. also test hand often after stopping

Reply:
I incorporated discussed changes and pushed the update.